src/app/tests/TestReadInteraction.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for CHIP Interaction Model Read Interaction
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <core/CHIPTLVUtilities.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <support/ErrorStr.h>
+#include <support/UnitTestRegistration.h>
+#include <system/SystemPacketBuffer.h>
+#include <system/TLVPacketBufferBackingStore.h>
+#include <transport/PASESession.h>
+#include <transport/SecureSessionMgr.h>
+#include <transport/raw/UDP.h>
+
+#include <nlunit-test.h>
+
+namespace chip {
+SecureSessionMgr gSessionManager;
+Messaging::ExchangeManager gExchangeManager;
+TransportMgr<Transport::UDP> gTransportManager;
+const Transport::AdminId gAdminId = 0;
+
+namespace app {
+class TestReadInteraction
+{
+public:
+    static void TestReadClient(nlTestSuite * apSuite, void * apContext);
+    static void TestReadHandler(nlTestSuite * apSuite, void * apContext);
+
+private:
+    static void GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle aPayload);
+};
+
+void TestReadInteraction::GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle aPayload)
```

Comment:
That last arg does not make sense.  We're not handing ownership over to this function... It should be `System::PacketBufferHandle & aPayload`, no?

Reply:
yep, agree, thanks for catching up it.